### PR TITLE
Add showOptOut option to spec

### PR DESCRIPTION
**Pull request by nickburris**
This PR adds spec text for a new showOptOut option in the SecurePaymentConfirmationRequest API. See the original proposal and discussion in #172. This spec change reflects the prototype currently implemented behind a flag in Chrome, which is available via origin trial and used in [this test page](https://rsolomakhin.github.io/pr/spc-opt-out/).

The OptOutError mentioned in this change is being added to WebIDL in https://github.com/whatwg/webidl/pull/1231.

Opt-out web platform test: https://github.com/web-platform-tests/wpt/pull/37012

***
<a href="https://pr-preview.s3.amazonaws.com/nickburris/secure-payment-confirmation/pull/215.html" title="Last updated on Dec 9, 2022, 1:49 PM UTC (78e2553)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/secure-payment-confirmation/215/06a08c0...nickburris:78e2553.html" title="Last updated on Dec 9, 2022, 1:49 PM UTC (78e2553)">Diff</a>

### CODE REVIEW

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:613]: We should not presume a specific UX (e.g., a button), and instead try to write for the general case.

This can be quite tricky, but you can see an example in https://w3c.github.io/secure-payment-confirmation/#sctn-transaction-confirmation-ux where we say today "the User Agent MUST ensure that the following is communicated to the user and that the user’s consent is collected for the authentication", rather than say (for example) "the User Agent MUST display a dialog with the following text fields".

For opt-out, I think the wording might be something like:

`"the user should be given a choice to opt-out during the [Transaction UX]"`

This comment applies to all references to 'button' below :)

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:777]: Here, instead of 'button is selected', we might use language something like: "If the user indicates that they wish to opt-out, the user agent must ..."

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:1521]: I think this is actually key text for this section, and so shouldn't be in a `NOTE`!

**Review by stephenmcgruer — CHANGES_REQUESTED** [review]
Thanks Nick! Some immediate comments about how we phrase this (sorry, specifications around UX are tricky :D)

**Comment by stephenmcgruer** [comment]
(cc @jcemer as an fyi)

File: spec.bs
Comment (nickburris) [reply spec.bs]: Done

File: spec.bs
Comment (nickburris) [reply spec.bs]: Done

File: spec.bs
Comment (nickburris) [reply spec.bs]: Done

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:780]: Nit; let's be a little more specific about what they opt out from:

'indicate that they want to opt out of the process for the given relying party'.

(Maybe linking given relying party to `rpId`?)

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:1518]:
```suggestion
opt out, an {{OptOutError}} is returned to the caller to indicate the user's
```

File: spec.bs
Comment (stephenmcgruer) [inline spec.bs:1522]:
```suggestion
This is not intended to be a mechanism to delete browser data or credentials - it
```

**Review by stephenmcgruer — APPROVED** [review]
Thanks Nick! I think this LGTM modulo a few nits.

Can we link to the WPT test PR (once available) in the PR description here?

**Comment by stephenmcgruer** [comment]
+@ianbjacobs to review as well. Ian - I expect we will want to talk about this one in the WPWG?

**Review by ianbjacobs — CHANGES_REQUESTED** [review]
Thank you for the pull request. I have a clarification request (must v. should) and suggest a privacy consideration addition. Thanks!

**Comment by ianbjacobs** [comment]
@stephenmcgruer, I will put this on the 8 December agenda.

File: spec.bs
Comment (nickburris) [reply spec.bs]: Done

**Comment by ianbjacobs** [comment]
Some day I will learn to use GitHub. In the meantime, here were the comments I thought I left:

* The proposal refers to this new feature using both "should" (twice) and "must." I recommend keeping the one "must" requirement and rewriting the other ones to be declarative, such as "provides the developer a way to ask the user agenda to provide the user with an opportunity to indicate they wish to opt out of the relying party's storage of information."
* I think the privacy considerations section should speak more directly to the fact that the opt out error lets the caller know the user has credentials, even though the user did not complete an authentication. I recommend including a statement about mechanisms that mitigate this (e.g., the user has taken explicit action via dedicated browser-owned UX). 
* Also, there is a "that that" in the proposal; drop one that. :)

**Comment by nickburris** [comment]
Thanks! See updates to the privacy considerations section.

> I think the privacy considerations section should speak more directly to the fact that the opt out error lets the caller know the user has credentials, even though the user did not complete an authentication. I recommend including a statement about mechanisms that mitigate this (e.g., the user has taken explicit action via dedicated browser-owned UX).

In our implementation we actually do not leak this information, since we also show the opt out option on the no-credential-found UX which already exists to mitigate credential probing attacks. I've added text to the privacy considerations section on this, good catch!

File: spec.bs
Comment (ianbjacobs) [inline spec.bs:1518]: Editorial simplification (now that I reread it):

"The API option {{SecurePaymentConfirmationRequest/showOptOut}} tells the user agent to provide a way for the user to indicate they wish to opt out of the relying party's storage of information."

**Review by ianbjacobs — APPROVED** [review]
See one more editorial suggestion. Thanks!

File: spec.bs
Comment (nickburris) [reply spec.bs]: Done

**Comment by nickburris** [comment]
Thanks again! As mentioned we'll leave this open until discussion at WPWG 8 December meeting.

**Comment by nickburris** [comment]
FYI I made a small change to update the naming convention in the TransactionAutomationMode enum to camelCase, per a recommendation in [this comment](https://chromium-review.googlesource.com/c/chromium/src/+/4030688/7/third_party/blink/public/devtools_protocol/browser_protocol.pdl#8126) which I'll follow up with an implementation change.

**Comment by stephenmcgruer** [comment]
@nickburris - fyi that this PR should also add a reference to the newly added WPT test (see https://github.com/w3c/secure-payment-confirmation/commit/c096e3b4ae4ca52b9c5d2e3cc9d850db64283751 where I landed it as `wpt hidden` for now; we should un-hide it and move it wherever in the spec seems most relevant!)

**Comment by stephenmcgruer** [comment]
This was discussed today at the WPWG ([minutes](https://www.w3.org/2022/12/08-wpwg-minutes.html#t01)), and a resolution was taken to land it (once updated for the WPT test comment). We also heard interest in exploring clearer cancellation states for SPC, which we intend to discuss internally - but consider separate from this PR.

**Comment by nickburris** [comment]
Sorry about the force push mess, one day I will get rebasing a fork right...

> @nickburris - fyi that this PR should also add a reference to the newly added WPT test (see [c096e3b](https://github.com/w3c/secure-payment-confirmation/commit/c096e3b4ae4ca52b9c5d2e3cc9d850db64283751) where I landed it as `wpt hidden` for now; we should un-hide it and move it wherever in the spec seems most relevant!)

Done

**Comment by stephenmcgruer** [comment]
> Sorry about the force push mess, one day I will get rebasing a fork right...

No problem! I also just rebased and forced-pushed it again, to pick up a fix I landed in `main`: https://github.com/w3c/secure-payment-confirmation/commit/06a08c0a82d63a74d699a7d8bb098f69d2f304cd

**Comment by stephenmcgruer** [comment]
@ianbjacobs - I think we're good to merge here, but will wait on you to confirm before I squash-merge it :)

**Pull request merged**
